Source/mpqapi.cpp

```diff
@@ -574,28 +574,20 @@ void CloseMPQ(const char *pszArchive, BOOL bFree, int dwChar)
 
 void mpqapi_store_modified_time(const char *pszArchive, int dwChar)
 {
-	int v2;                                // esi
-	const char *v3;                        // ebx
-	HANDLE v4;                             // eax
-	int v5;                                // esi
-	struct _WIN32_FIND_DATAA FindFileData; // [esp+8h] [ebp-1E0h]
-	char dst[160];                         // [esp+148h] [ebp-A0h]
+	HANDLE handle;
+	struct _WIN32_FIND_DATAA FindFileData;
+	char dst[160];
 
-	v2 = dwChar;
-	v3 = pszArchive;
 	if (gbMaxPlayers != 1) {
 		mpqapi_reg_load_modification_time(dst, 160);
-		v4 = FindFirstFile(v3, &FindFileData);
-		if (v4 != INVALID_HANDLE_VALUE) {
-			FindClose(v4);
-			v5 = 16 * v2;
-			*(_DWORD *)&dst[v5 + 8] = FindFileData.ftLastWriteTime.dwLowDateTime;
-			*(_DWORD *)&dst[v5 + 12] = FindFileData.ftLastWriteTime.dwHighDateTime;
+		handle = FindFirstFile(pszArchive, &FindFileData);
+		if (handle != INVALID_HANDLE_VALUE) {
+			FindClose(handle);
+			*((FILETIME*) (dst) + dwChar * 2 + 1) = FindFileData.ftLastWriteTime;
 			mpqapi_reg_store_modification_time(dst, 160);

```

Comment:
This looks very suspicious to me.  Why does the offset of the destination of `ftLastWriteTime` being copied depends on the count of characters or whatever `dwChar` is?

Reply:
If we look a bit higher up the stack `dwChar` is actually named `save_num` and there it basically player number. So I think it just ends up as 4 possible pairs of modification/creation time. I think renaming arguments etc. could happen separate from clean up since I have very little idea what is going on in this file as a whole to be honest.

Reply:
I think this is used as a "save file modification protection". So that the time stamps of the save files is stored by `Diablo.exe` in the regedit when running the game. If a user were to modify the save file manually (i.e. when not running `Diablo.exe`) then the modification time of the actual save file `foo.sv` would not match up with those in regedit. At least, that's my understanding of the functionality of this code. Also, the reg key is stored as "Video Player" something, just to avoid suspicion for people trying to hack the game (and who are actively looking thorough the regedit to see what's there).

Reply:
I guess this part of the code was eventually disabled since the game no longer saves the value and you can freely move save games between machines.

Reply:
Yeah. Oh, and just to confirm. I don't know if "save file modification protection" is what it was actually used for. That's just what made most sense to me, considering it's storing the time stamps and obfuscating the name of the registry key.

Reply:
This probably explains the function `mpqapi_store_default_time` and why it was active in the debug release but removed from the final game.